api/src/trace.rs

```diff
@@ -63,3 +64,72 @@ pub enum ExecutionEvent {
         msg: String,
     },
 }
+
+impl From<ExecutionTrace> for InvocationTrace {
+    fn from(e_trace: ExecutionTrace) -> InvocationTrace {
+        let mut invocation_stack: Vec<InvocationTrace> = Vec::new();
+
+        for event in e_trace.clone().events.into_iter() {
+            match event {
+                ExecutionEvent::GasCharge { .. } | ExecutionEvent::Log { .. } => {}
+                ExecutionEvent::Call { from, to, method, params, value } => {
+                    invocation_stack.push(InvocationTrace {
+                        from: Address::new_id(from),
+                        to,
+                        method,
+                        params,
+                        value,
+                        code: ExitCode::OK, // Placeholder, will be updated during return
+                        ret: None,          // Placeholder, will be updated during return
+                        subinvocations: Vec::new(),
+                    });
+                }
+                ExecutionEvent::CallReturn { return_value, exit_code } => {
+                    let mut current_invocation = invocation_stack
+                        .pop()
+                        .unwrap_or_else(|| panic!("Unmatched call return: {:?}", e_trace));
+
+                    current_invocation.code = exit_code;
+                    current_invocation.ret = return_value;
+
+                    if let Some(parent_invocation) = invocation_stack.last_mut() {
+                        parent_invocation.subinvocations.push(current_invocation);
+                    } else {
+                        // At root invocation
+                        return current_invocation;
+                    }
+                }
+                ExecutionEvent::CallAbort { exit_code } => {
+                    let mut current_invocation = invocation_stack
+                        .pop()
+                        .unwrap_or_else(|| panic!("Unmatched call abort: {:?}", e_trace));
+
+                    current_invocation.code = exit_code;
+
+                    if let Some(parent_invocation) = invocation_stack.last_mut() {
+                        parent_invocation.subinvocations.push(current_invocation);
+                    } else {
+                        // At root invocation
+                        return current_invocation;
+                    }
+                }
+                ExecutionEvent::CallError { .. } => {
+                    let mut current_invocation = invocation_stack
+                        .pop()
+                        .unwrap_or_else(|| panic!("Unmatched call error: {:?}", e_trace));
+
+                    current_invocation.code = ExitCode::SYS_ASSERTION_FAILED;
```

Comment:
Is this the right ExitCode or is there a defined mapping from ErrorNumber to ExitCode?

Should the error message or error number be surfaced in the trace somehow?

Reply:
Yes, there is a defined mapping in actors builtin/shared.rs `SendError`, it's not a property of the VM. This is a mismatch in abstraction level. Probably we should change the InvocationTrace to be able to represent this error number xor an exit code. But assertion failed is good for most instances for now.